Made changes to code pipeline to allow main.py to run. Also fixed several TODO's

dickeygh opened the pull request:
### RAW COMMITS:
COMMIT 1: 
Added "from core.etc..." to lines 17-23 of models.py.
Uncommented the confusion_matrix and classification_report on lines 101-102 of train.py


COMMIT 2:
 Added "from core import...." to lines 9-11 of compare.py so that the …
…code would compile/run properly.
Added "import cirpy" to ingest.py so that line 53 could run, as it uses cirpy.


COMMIT 3:
 Fixed the "TODO Make task identification automatic based on dataset" …
…by adding an if statement that sets self.task_type to 'regression' or 'classification' based on the dataset that was being used.
Also removed "task" from the init variables as a result.


COMMIT 4:
 IN MODELS.PY: Added features.featurize and features.datasplit functio…
…ns to the run() function in order to featurize and split the dataset. Commented out lines 118-120, line 128, line 144, line 147, and line 153, as main.py would not run properly with these lines.
IN NAME.PY: Changed the file directory to match my local directory, as I could not get it to run with any version of "../dataFiles" etc.
IN MAIN.PY: Removed the drop=True argument in line 66, as this is no longer necessary. Commented out line 75, as model.featurization is no longer a function. Removed the tune=True argument in line 77, as this is no longer necessary.


COMMIT 5:
 IN MODELS.PY: Changed lines 158-163 to be correctly commented for my …
…windows machine. Could we just ask the user what os they are using?
Added comments to lines 82 and 83.
IN MAIN.PY: Deleted un-necessary old comments/code.


COMMIT 6:
 IN MODELS.PY: Changed feat_meth to receive a featurization option num…
…ber (from main.py) instead of using [0] every time.
IN NAME.PY: Added classifier models to algorithm_list
IN TRAIN.PY: Commented out lines 79 and 80 to use conf and clsrep as individual variables instead of arrays. Added print statements to show the user accuracy score, confusion matrix, classification report, and roc_auc_score for each run. Commented out the .mean and .std for classification report, as it is attempting to average string variables (lines 125-130).
IN MAIN.PY: Changed line 66 and line 90 to send a method (featurization option). Removed out-dated code at old lines 97-98. Changed line 98 to fit new pipeline.


COMMIT 7:
IN MAIN.PY: Added comments next to the declaration of feats list. Rem…
…oved un-necessary code at lines 51-54.



### SUMMARY:

**COMPARE.PY**
1. Changed lines 9-11 to have "from core" so that it would run properly.
2. Commented out line 444 so it would run properly.


**INGEST.PY**
1. Added "import cirpy" at line 3 in order to have it run properly.


**MODELS.PY**
1. Added "from core." to lines 17-23 so it would run properly.
2. Removed task declaration (line 27). This is because the program now automatically sets self.task_type based on the data set being used (lines 34-38)
3. Changed feat_meth = [0] to feat_meth (line 27). This is so it will use whichever featurization option is set in main.py.
4. Added featurization and datasplit to init function (lines 82-83)
5. Commented out lines 117-120, 128, 144, 153 because they were breaking it while running from main.py.
6. Changed comments at lines 158-163 to work on my windows machine.


**NAME.PY**
1. Changed directory in line 17 to work on my machine (could not get it to work with any variation of ./dataFiles for some reason).
2. Added classification algorithms to algorithm_list (line 68). This is so classification would run properly from main.py.


**TRAIN.PY**
1. Commented out lines 79-80 so that conf and clsrep could be used as non-array variables.
2. Changed lines 100-112 to print out classification metrics for each run.
3. Changed comments from lines 125-130 in order to get it to run properly. This is because some of these lines are trying to average string variables. I will look for a way around this so that we can use arrays to store the classification report and confusion matrix.


**MAIN.PY**
1. Added comments to lines 26 and 39 (this is where you set up which featurization options you want). 
2. Removed un-necessary code from lines 51-54.
3. Changed line 62 to send the featurization method and removed drop=True to match new pipeline.
4. Removed un-necessary code from old lines 74-75.
5. Removed tune=True from line 72 to fit new pipeline.
6. Added model.analyze at line 73.
7. Changed line 86 to send a featurization method and removed drop=False to match new pipeline.
8. Removed un-necessary code from old lines 97-98.
9. Changed line 94 to model.run() instead of model.classification_run(). This fits the new pipeline.


**Other notes/thoughts:** I apologize if any of my solutions or changes are not taking the pipeline in the direction that has been envisioned. I left all of the TODO comments in the code in case the solutions that I implemented are not accomplishing the perceived goal. I envision this PR going through multiple iterations before it is ready to be sent to the dev branch. Please check this out, run main.py, and provide me with feedback.

pep8speaks commented:
Hello @dickeygh! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`core/compare.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/compare.py):

> [Line 444:1](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/compare.py#L444): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '



* In the file [`core/models.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py):

> [Line 17:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 27:5](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L27): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 27:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L27): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 79 characters)
> [Line 27:101](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L27): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 27:103](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L27): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 37:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L37): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (189 > 79 characters)
> [Line 82:33](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L82): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 83:34](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L83): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 117:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L117): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 118:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L118): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 119:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L119): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 120:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L120): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 128:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L128): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 144:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L144): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 147:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L147): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 147:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L147): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 153:9](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L153): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 153:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L153): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 158:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L158): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 159:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L159): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)
> [Line 162:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L162): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 163:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/models.py#L163): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)

* In the file [`core/name.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/name.py):

> [Line 17:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/name.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 79 characters)
> [Line 68:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/name.py#L68): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 79 characters)

* In the file [`core/train.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/train.py):

> [Line 79:5](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/train.py#L79): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 80:5](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/train.py#L80): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 129:8](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/core/train.py#L129): [E131](https://duckduckgo.com/?q=pep8%20E131) continuation line unaligned for hanging indent

* In the file [`main.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py):

> [Line 26:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py#L26): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (169 > 79 characters)
> [Line 26:87](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py#L26): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 39:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (165 > 79 characters)
> [Line 39:87](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py#L39): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 73:36](https://github.com/Dlux804/McQuade-Chem-ML/blob/b0992ddc004946a6ec85fb0576897eb94228c2e0/main.py#L73): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

andreshyer commented on core/name.py:
```diff
@@ -14,7 +14,7 @@ def represent_dataset():
     :return: Dictionary of dataset as key and their representation as value
     """
-    with misc.cd('../dataFiles'):  # Access folder with all dataset
+    with misc.cd(r'C:\Users\brend\PycharmProjects\McQuade-Chem-ML4\dataFiles'):  # Access folder with all dataset
```
This will not work for other people, only on your machine. Try something like
`with misc.cd(str(pathlib.Path(__file__).parent.parent.absolute()) + '/dataFiles/'):` which looks for the absoulate path for any machine. I just ran into this bug myself

dickeygh replied:
This seems to have worked. Will include this change in the next iteration of this PR. Thanks!

andreshyer commented on main.py:
```diff
@@ -48,10 +48,6 @@ def main():
-        if alg == 'svc':
-            feats = [[0], [0, 1], [0, 2], [0, 3], [0, 4], [1], [2], [3], [4]] # Removes atompaircounts featurization only when running svc, as this featurization breaks the svc model.
-        elif alg != 'svc':
-            feats = [[0], [0, 1], [0, 2], [0, 3], [0, 4], [0, 5], [1], [2], [3], [4], [5]] # Insures that featurization options don't change for models running after svc in iteration.
```
Is this safe to remove?

dickeygh replied:
This should be safe to remove as this was based on an older version of `features.py`.

dickeygh replied:
I envision changing main.py to have different 'feats' definitions based on which model is being ran. These definitions will change from previously (what was removed here) because features.py no longer removes a featurization option based on the model. Does this make sense?

andreshyer replied:
I think so, I am completely fimilar with the how the different feats interact with the different models, but what you are saying does make sense.

andreshyer commented on main.py:
```diff
@@ -87,17 +83,15 @@ def main():
                     print('Dataset:', data)
                     print()
-                    # featurize molecules
-                    model.featurization(method)
```
it is still important you featurize the data. Otherwise the models do not make sense to run. So instead of removing this line, change it to `model.featurize()`

dickeygh replied:
This is currently being done in the models.py run() function (line 82). Is there an issue with using this approach?

andreshyer commented on core/models.py:
```diff
@@ -72,6 +79,8 @@ def run(self):
             self.make_grid()
             self.hyperTune(n_jobs =6)
 
+        features.featurize(self) # Featurize Molecules
```
Now I see you do featurize the model inside of models.py. Yeah, I would not recommend this. We would ideally like to featurize models once, then save those features (maybe in a json file) to be used again later. So I would highly recommend leaving this in the main.py

dickeygh replied:
Ok, I will do this. Thank you!

qle2 commented on core/models.py:
```diff
@@ -14,23 +14,30 @@ class MlModel:  # TODO update documentation here
+        # Sets self.task_type based on which dataset is being used.
+        if self.dataset == 'sider.csv':
+            self.task_type = 'classification'
+        elif self.dataset == 'Lipophilicity-ID.csv' or self.dataset == 'ESOL.csv' or self.dataset == 'water-energy.csv' or self.dataset == 'logP14k.csv' or self.dataset == 'jak2_pic50.csv':
```
Change this to `elif self.dataset in ['Lipophilicity-ID.csv','ESOL.csv', ... ]`. It looks way cleaner that way.

dickeygh commented:
In the last commit, I made the following changes:

**MODELS.PY**
1. Removed line 82 because this was moved to main.py (lines 72 and 95)


**NAME.PY**
1. Added import pathlib (line 4) so that it could be used in line 17.


**MAIN.PY**
1. Added "from core.features import featurize" (line 7) so that featurization could be performed in main.py (lines 72 and 95).

qle2 commented on core/models.py:
```diff
@@ -14,23 +14,30 @@ class MlModel:  # TODO update documentation here
         self.algorithm = algorithm
         self.dataset = dataset
+
+        # Sets self.task_type based on which dataset is being used.
+        if self.dataset == 'sider.csv':
+            self.task_type = 'classification'
+        elif self.dataset == 'Lipophilicity-ID.csv' or self.dataset == 'ESOL.csv' or self.dataset == 'water-energy.csv' or self.dataset == 'logP14k.csv' or self.dataset == 'jak2_pic50.csv':
+            self.task_type = 'regression'
```
While I understand the logic behind it, this is too much hard-coding. Since we will have more classification dataset in the future assuming @dickeygh continues his work in classification, we will probably need separate folders for classification data and regression data. Then we can directly check to see which type of algorithm `self.dataset` requires based on where it's located instead of hard-coding it in. What do you think about this idea @andreshyer ?

Dlux804 replied:
I agree, this method of hard coding is not the way to go.  At the very least, a dict or list would be better for checking.  `if self.dataset is in reg_list: ...` or something would be preferred.  A more sophisticated method would be to look at the supplied target column and determine if it is continuous or discrete.

andreshyer replied:
Yeah, I agree with Adam. I believe a dict of would be a more append-able way of telling the difference between regression and classification

andreshyer replied:
Or perhaps it you wanted to completely automated it, test the column data type. (Int/strings for discrete, and float for continuous)

dickeygh replied:
Made a list of all the regression datasets to check `self.dataset` against as a temporary solution.

qle2 commented on core/models.py:
```diff
@@ -105,18 +113,18 @@ def store(self):
         # create dictionary of attributes
         att = dict(vars(self))  # makes copy so does not affect original attributes
         del att['data']  # do not want DF in dict
-        del att['smiles']  # do not want series in dict
-        del att['graphM']  # do not want graph object
-        del att['stats']  # will unpack and add on
-        del att['pvaM']  # do not want DF in dict
+        #del att['smiles']  # do not want series in dict
+        #del att['graphM']  # do not want graph object
+        #del att['stats']  # will unpack and add on
+        #del att['pvaM']  # do not want DF in dict
```
These variables are removed because we don't need to record them for our results and the way they are unpacked in a CSV is hedious. Do you see we use them `graphM`, `smiles`, `stats` and `pvaM` in anyway @dickeygh ?

dickeygh replied:
When these are uncommented, main.py breaks when trying to run model.store() for the regression pipeline.

qle2 commented on core/models.py:
```diff
@@ -105,18 +113,18 @@ def store(self):
             pass
         # del att['impgraph']
-        att.update(self.stats)
+        #att.update(self.stats)
```
Why did you want to remove this line?

dickeygh replied:
When this line is uncommented, main.py breaks when trying to run model.store() for the regression pipeline. Error message: `AttributeError: 'MlModel' object has no attribute 'stats'`

qle2 commented on core/models.py:
```diff
@@ -132,26 +140,26 @@ def store(self):
-        self.graphM.close()  # close to conserve memory when running many models.
+        #self.graphM.close()  # close to conserve memory when running many models.
         # self.graph.savefig(name+'PvA')
 
         # make folders for each run
         # put output files into new folder
-        # filesp = ''.join(['move ./', self.run_name, '* ', self.run_name, '/'])  # move for Windows system
-        filesp = ''.join(['mv ./', self.run_name, '* ', self.run_name, '/'])  # mv for Linux system
+        filesp = ''.join(['move ./', self.run_name, '* ', self.run_name, '/'])  # move for Windows system
+        # filesp = ''.join(['mv ./', self.run_name, '* ', self.run_name, '/'])  # mv for Linux system
         subprocess.Popen(filesp, shell=True, stdout=subprocess.PIPE)  # run bash command
 
-        # movepkl = ''.join(['move ./', '.pkl', '* ', self.run_name, '/'])  # move for Windows system
-        movepkl = ''.join(['mv ./', '.pkl', '* ', self.run_name, '/']) # mv for Linux system
+        movepkl = ''.join(['move ./', '.pkl', '* ', self.run_name, '/'])  # move for Windows system
+        # movepkl = ''.join(['mv ./', '.pkl', '* ', self.run_name, '/']) # mv for Linux system
```
Instead of switching commands between Linux and Windows, I think it's better if we create another method to move CSV files that is independent of operating systems.

dickeygh replied:
Could asking the user which OS they are in or hard coding this somewhere in `main.py` be a temporary solution?
